Rewrite docs for authenticated feeds (#3201)
NuGet/docs.microsoft.com-nuget: docs/consume-packages/consuming-packages-authenticated-feeds.md

description: Consuming packages from authenticated feeds in all NuGet client scenarios
author: nkolev92
ms.author: nikolev
ms.date: 12/22/2023
ms.topic: conceptual
---

# Consuming packages from authenticated feeds

Many NuGet operations, such as restore and install, require communication with one or more package sources, which [can be configured in *nuget.config* files](../reference/nuget-config-file.md#packagesources).
For HTTP feeds, NuGet will with make an unauthenticated request, and if the server responds with an HTTP 401 response, NuGet will search for credentials in the following order:

1. [An environment variable `NuGetPackageSourceCredentials_{name}`](#credentials-in-environment-variables).

> 1. [Add credentials in a *nuget.config* file](#credentials-in-nugetconfig-files).

Reviewer: ```suggestion
1. [Credentials in the *nuget.config* files](#credentials-in-nugetconfig-files).
```

1. [Use a NuGet credential provider, if your package source provides one](#credential-providers).

> We recommend using a credential provider when possible.

Reviewer: I feel like this should have a larger emphasis.

Reviewer 2: > [!NOTE]
> We recommend using a credential provider when possible.

> Secrets wil not be saved in the *nuget.config* file, reducing risk of accidentally leaking secrets.

Reviewer 3: ```suggestion
Secrets will not be saved in the *nuget.config* file, reducing risk of accidentally leaking secrets.
```

description: Consuming packages from authenticated feeds in all NuGet client scenarios
author: nkolev92
ms.author: nikolev
ms.date: 12/22/2023
ms.topic: conceptual
---

# Consuming packages from authenticated feeds

Many NuGet operations, such as restore and install, require communication with one or more package sources, which [can be configured in *nuget.config* files](../reference/nuget-config-file.md#packagesources).
For HTTP feeds, NuGet will with make an unauthenticated request, and if the server responds with an HTTP 401 response, NuGet will search for credentials in the following order:

1. [An environment variable `NuGetPackageSourceCredentials_{name}`](#credentials-in-environment-variables).
1. [Add credentials in a *nuget.config* file](#credentials-in-nugetconfig-files).
1. [Use a NuGet credential provider, if your package source provides one](#credential-providers).

> 1. [Interactive login](#interactive-login)

Author: If no cred provider provides a credential for a particular source (for example, try a github package registry feed), then NuGet itself will prompt for a username and password (when in interactive mode).  This is what this section is referring to.

> 1. [Interactive login](#interactive-login)

Reviewer 4: > I also thought that NuGet in VS would pop up an interactive username/password box when everything else fails, but I can neither find the code or reproduce it.

I was able to reproduce interactive username/password box on a GitHub package registry.

description: Consuming packages from authenticated feeds in all NuGet client scenarios
author: nkolev92
ms.author: nikolev
ms.date: 12/22/2023
ms.topic: conceptual
---

# Consuming packages from authenticated feeds

Many NuGet operations, such as restore and install, require communication with one or more package sources, which [can be configured in *nuget.config* files](../reference/nuget-config-file.md#packagesources).
For HTTP feeds, NuGet will with make an unauthenticated request, and if the server responds with an HTTP 401 response, NuGet will search for credentials in the following order:

1. [An environment variable `NuGetPackageSourceCredentials_{name}`](#credentials-in-environment-variables).
1. [Credentials in *nuget.config* files](#credentials-in-nugetconfig-files).
1. [Use a NuGet credential provider, if your package source provides one](#credential-providers).

> [!NOTE]
> We recommend using a credential provider when possible.

> > Secrets will not be saved in the *nuget.config* file, reducing risk of accidentally leaking secrets.

Reviewer 2: ```suggestion
> Secrets should not be saved in the *nuget.config* file, reducing risk of accidentally leaking secrets.

```

> Additionally, it typically reduces the number of places you need to update when a credential expires or changes.

> > If the credential provider supports single sign on, it may reduce the number of time you need to log in, or the number of places that credentials need to be saved.

Reviewer 2: ```suggestion
> If the credential provider supports single sign-on, it may reduce the number of times you need to login, or the number of places that credentials need to be saved.

```

The credentials you need to use are determined by the package source.

> Therefore, unless you're using a credential provider, you should check with your package source what credentials to use.

Reviewer 2: ```suggestion
Therefore, unless you're using a credential provider, you should check with your package source for what credentials to use.

```

> It is very common for package sources to forbid using your password (that you log into the website with) with NuGet.

Reviewer 2: ```suggestion
It is very common for package sources to forbid you from using your password (that you log into the website with) with NuGet.

```

Typically you need to create a Personal Access Token to use as NuGet's password, but you should check the documentation for the NuGet server you're using.
Some package sources, such as Azure DevOps and GitHub, have scoped access tokens, so you may need to ensure that any tokens you create include the required scope.

> ## Credentials in environment variables

Reviewer: I've always found environment variables as configurations to really reduce transparency since they are usually the highest priority in any config mode, and it's really difficult to know where something is configured.

I think that should ideally be represented, but I'd love to hear opinions.

For example, consider the following *nuget.config* file:

```xml
<configuration>
  <packageSources>
    <clear />
    <add key="Contoso" value="https://nuget.contoso.com/v3/index.json" />
  </packageSources>
</configuration>
```

In this case, the source name is `Contoso` and NuGet will look for the environment variable name `NuGetPackageSourceCredentials_Contoso`.
Some platforms are case-sensitive, so take care about using the correct upper and lower case characters for the environment name and the source name, as defined in your *nuget.config* file.

If the username is `nugetUser` and the password is `secret123`, the environment variable's value should be set to `Username=nugetUser;Password=secret123`.
If NuGet should only use this credential for HTTP Basic authentication, but not other authentication schemes, you can set the environment variable's value to `Username=nugetUser;Password=secret123;ValidAuthenticationTypes=Basic`.
For more information about valid authentication types, see [the docs on package credentials in *nuget.config* files](../reference/nuget-config-file.md#packagesourcecredentials).

> Note that environment variables have restrictions on allowed characters, and different operating systems may have different restrictions.

Reviewer 2: ```suggestion
> [!NOTE] 
> Environment variables have restrictions on allowed characters, and different operating systems may have different restrictions.

```

> For example, spaces are not allowed.

Reviewer 2: Include in note above
